File: web_version/pmugg_files/guidelines/faceGroup.js
// Define a group of faces. The first face is the outer component. The
// other faces are inner components and represent holes in the face.
ms.faceGroup = function(stats) {
	var properties = {
		face: new ms.alternativeArray('face', true),
	};
	this.node = new ms.node(this, stats, 'faceGroup', properties);
};

ms.faceGroup.prototype.getNode = function() {
	return this.node;
};

ms.faceGroup.prototype.getFaces = function() {
	return this.node.get('face');
};

// Connect a new face as an inner component or hole to this one.
ms.faceGroup.prototype.connectHole = function(groupB) {
	if (groupB == this) {
		// This face has already been added.
		return;
	}
	var facesB = groupB.getFaces();
	if (facesB.length != 1) {
		ms.alert('Hole group should have one face');
	}
	var faceB = facesB[0];
	faceB.getNode().disconnect(groupB);
	faceB.getNode().connect(this);
};

// ms.faceGroup.prototype.print = function() {	ms.highlight(this); };

File: cpp_version/graph_drawing/face_group.cpp
#include "pch.h"
#include "face_group.h"

namespace ms {

FaceGroup::FaceGroup(Model* model, int id) : model(model), id(id), faceIds() {
    model->getCurrent()->addFaceGroup(id, this);
}

FaceGroup::FaceGroup(Model* model, int id, ::vector<int> faceIds) : model(model), id(id), faceIds(faceIds) {
    model->getCurrent()->addFaceGroup(id, this);
}

FaceGroup* FaceGroup::copy() {
	return new FaceGroup(model, id, faceIds);
}

void FaceGroup::addFace(Face* face) {
    faceIds.push_back(face->getId());
    face->setGroup(this);
}

void FaceGroup::insertFace(Face* face, int index) {
    faceIds.insert(faceIds.begin() + index, face->getId());
    face->setGroup(this);
}

void FaceGroup::removeFace(Face* face) {
    faceIds.erase(std::remove(faceIds.begin(), faceIds.end(), face->getId()), faceIds.end());
}

::vector<Face*> FaceGroup::getFaces() const {
	::vector<Face*> result;
	for (int i = 0; i < faceIds.size(); i++) {
		result.push_back(model->getCurrent()->getFace(faceIds[i]));
	}
	return result;
}

void FaceGroup::destroyIfEmpty() {
    if (faceIds.size() == 0) {
        model->getCurrent()->removeFaceGroup(this);
        delete this;
    }
}

void FaceGroup::connectHole(FaceGroup* groupB) {
    if (groupB == this) {
        // This face has already been added.
        return;
    }
    auto facesB = groupB->getFaces();
	if (facesB.size() != 1) {
        ::cout << "Hole group should have one face" << ::endl;
	}
	auto faceB = facesB[0];
    groupB->removeFace(faceB);
    addFace(faceB);
}

} // namespace ms 